### PR TITLE
Implement Read, BufRead, Write and Seek for Cursor<Box<[u8]>>

### PR DESCRIPTION
**Pull request by sfackler**

### CODE REVIEW

File: src/libstd/io/cursor.rs [hunk 1]
```diff
@@ -289,6 +295,16 @@ impl Write for Cursor<Vec<u8>> {
     fn flush(&mut self) -> io::Result<()> { Ok(()) }
 }
 
+#[stable(feature = "cursor_box_slice", since = "1.4.0")]
```
Comment (bluss) [inline src/libstd/io/cursor.rs]: There could be a doc comment here explaining the semantics. It *should* be visible on the type's rustdoc page in the current implementation (so on Cursor's page).

The vector growing and the box merely filling is something that we probably want to put into the documentation somehow.

Comment (sfackler) [inline src/libstd/io/cursor.rs]: Agreed - I have personally been a bit confused as to the exact semantics of some of the impls here. I believe that docs on the impl itself will not be rendered but docs on methods in the impl will.

Comment (bluss) [inline src/libstd/io/cursor.rs]: It's actually the opposite in my experience. #24838

**Comment by alexcrichton** [comment]
I'm a little worried about the interactions between this and #27197, as I think that attempting to generalize these implementations is probably a good idea and don't want to gain too many impl blocks here. I think that using the generalization of `Deref` is a bit too powerful (and a bit unidiomatic), and as found in #27197 `AsRef` won't work as it's not implemented for smart pointers. Perhaps, though, the `BufRead`, `Read`, and `Seek` implementations could be generalized for `T: Borrow<[u8]>`? That's implemented for all smart pointer types and would give us `Cursor<Box<[u8]>>` for free (reading at least).

For writing, we could generalize over `T: BorrowMut<[u8]>` to get the `&mut [u8]` and `Box<[u8]>` implementations, but it would invalidate the current `Vec<u8>` implementation (e.g. it wouldn't allow the vector to grow). With specialization, however, we should be able to have this overlap.

In that sense I would be ok exploring using `Borrow` for the reading side of things and adding an explicit impl of `Write for Cursor<Box<[u8]>>`.  How does that sound?

cc @nwin

**Comment by nwin** [comment]
I agree that the write part can’t be generalized easily, so I think it would be ok to implement Cursor for Box<[u8]>.

I'm not sure about the `Borrow`-part, as already discussed in #23364, it would not work with fixed-sized arrays. Furthermore `Borrow` has been implemented for `Box` only very recently (not in May at least) so it this really the appropriate `trait` for this task? I would be in favor of using `AsRef` and forward-implement it for all smart pointers as this would cover all slice-like structures I can think of.

**Comment by alexcrichton** [comment]
@nwin unfortunately `AsRef` is off the table due to the breakage of `as_ref` on  `Box<Option<T>>`

**Comment by nwin** [comment]
@alexcrichton: That is really unfortunate because then I do not see a nice solution.

Seems we have the following options:

 - Implement `Cursor` for `Box<[T]>` without forgetting `Rc<[T]>`, `Arc<[T]>` and reminding everybody to also implement it for any custom smart pointer aka. `Gc<[T]>`
 - Use `Deref<Target=AsRef<[u8]>>` and not care about raw `[T; N]` (`Cursor::new([0; 2])`)
 - Use `Borrow` and not care about any `[T; N]`

**Comment by sfackler** [comment]
None of `Rc<[T]>`, `Arc<[T]>` and `Gc<[T]>` can be used in a Cursor since they don't return mutable references to their interior.

**Comment by sfackler** [comment]
Er, that only applies to `Write` impls, sorry.

**Comment by alexcrichton** [comment]
Given https://github.com/rust-lang/rust/pull/27197 (which I'd personally be ok accepting) that would alter this to just adding `Write` for `Box<[u8]>`, which I think is fine. Let's wait to see what happens over there though.

**Comment by alexcrichton** [comment]
The libs team is still holding off on this until #27197 is resolved (as a status update)

**Comment by alexcrichton** [comment]
@sfackler now that https://github.com/rust-lang/rust/pull/28811 has landed which has unblocked https://github.com/rust-lang/rust/pull/27197 which will in turn unblock this, just a heads up that this may need a rebase soon! I think that the parts we agreed to add were `Write for Cursor<Box<[u8]>>` (having the same semantics as `&mut [u8]`, and that should do the trick!

**Comment by sfackler** [comment]
rebased

File: src/libstd/io/cursor.rs (hunk 1 above)
Comment (alexcrichton) [inline src/libstd/io/cursor.rs]: Ah this should actually be 1.5.0 now (oh how time flies!)

**Comment by alexcrichton** [comment]
r=me with minor nit about the stable since

**Pull request merged**